src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs

```diff
@@ -72,7 +72,13 @@ internal class DefaultHealthCheckService : HealthCheckService
                         HealthReportEntry entry;
                         try
                         {
-                            var result = await healthCheck.CheckHealthAsync(context, cancellationToken);
+                            HealthCheckResult result;
+                            using (var timeoutCancellationTokenSource = new CancellationTokenSource(registration.Timeout))
+                            using (cancellationToken.Register(() => timeoutCancellationTokenSource.Cancel()))
+                            {
+                                result = await healthCheck.CheckHealthAsync(context, timeoutCancellationTokenSource.Token);
+                            }
+
```

Comment:
Can we find a way to make the allocation of `timeoutCancellationTokenSource` skip when the timeout is infinite? I suspect an infinite timeout is going to be the common case.

Reply:
Done